### PR TITLE
Improving Python DataType support for Struct and repr

### PR DESCRIPTION
**Pull request by cjermain**
While working on #3413, I noticed that the Python representation of DataTypes has some issues. This PR fixes a few of them:

- Struct dtype could not be converted back to `DataType::Struct`, since previously it had an incomplete representation
- `dtype_string_repr` was not implemented yet, so it always failed to import

The `__repr__` behavior is a bit odd, but much closer in this PR to the Rust API. This is because the class is being used as a singleton without being constructed -- `type(dtype) == type`. So for native types `repr(dtype)` is actually `type(dtype).__repr__(dtype)` which gives strings like `"<class 'polars.datatypes.Utf8'>"`. I left the `DataType.string_repr` method, but made it more clear that as a classmethod this uses `cls` not `self`. I think this could be improved by having an instantiated singleton that the class provides on the new constructor. This can be a follow up.

### CODE REVIEW

**Comment by cjermain** [comment]
I'm not seeing the tests fail locally. I'm wondering if the `test_arr_unique` test is intermittent:
https://github.com/pola-rs/polars/runs/6543827625?check_suite_focus=true#step:9:514

**Comment by ritchie46** [comment]
Could you do a rebase on master? That should fix it.

File: py-polars/src/conversion.rs [hunk 1]
```diff
@@ -308,57 +311,65 @@ impl FromPyObject<'_> for Wrap<QuantileInterpolOptions> {
...
-            "Datetime" => DataType::Datetime(TimeUnit::Microseconds, None),
-            "Time" => DataType::Time,
-            "Duration" => DataType::Duration(TimeUnit::Microseconds),
-            "Float32" => DataType::Float32,
-            "Float64" => DataType::Float64,
-            "Object" => DataType::Object("unknown"),
-            // just the class, not an object
-            "List" => DataType::List(Box::new(DataType::Boolean)),
-            "Null" => DataType::Null,
+        let type_name = ob.get_type().name()?;
```
Comment (ritchie46) [inline py-polars/src/conversion.rs:324]: Much better!

**Comment by ritchie46** [comment]
Thanks for this. I agree it's something we can improve because of the nested dtypes.

**Pull request merged**